common/web/gesture-recognizer/src/pathSegmenter.ts

```diff
@@ -0,0 +1,957 @@
+                choppedStats: CumulativePathStats,
+                splitIndex: number) {
+      const pre        = steppedStats[splitIndex].deaccumulate(choppedStats);
+
+      // Keep stats value components based on the final point of the 'pre' segment.
+      const finalStats = steppedStats[steppedStats.length-1];
+      const post       = finalStats.deaccumulate(steppedStats[splitIndex-1]);
+      const union      = finalStats.deaccumulate(choppedStats);
+
+      super(pre, post, union, finalStats);
+      this.baseChop  = choppedStats;
+      this.chopPoint = steppedStats[splitIndex-1];
+      this.endOfPre  = steppedStats[splitIndex];
+    }
+  }
+
+  /**
+   * The core logic, algorithm, and manager for touchpath segmentation.
+   */
+  export class PathSegmenter {
+    /**
+     * The minimum amount of time (in ms) to wait between sample repetitions
+     * once inputs stop arriving, so long as the path is still active.
+     */
+    private readonly REPEAT_INTERVAL = 33;
+
+    /**
+     * The time-interval length (in ms) at the end of the path to consider
+     * when determining whether or not to trigger path segmentation.
+     */
+    private readonly SLIDING_WINDOW_INTERVAL = 50;
+
+    /**
+     * Tracks the mathematical values used to provide path segment stats
+     * at each point on the path.  Individual steps may be removed (and
+     * batched into `choppedStats`) once their respective points on the
+     * path have been fully processed.
+     */
+    private steppedCumulativeStats: CumulativePathStats[];
+
+    /**
+     * Tracks all subsegments awaiting completion of their overall segment.
+     */
+    private lingeringSubsegmentations: PotentialSegmentation[];
+
+    /**
+     * TODO: These directly represent path segments produced by the prototype
+     * algorithm.  Just... the stats analysis of the path segment, without
+     * obvious / public members to relevant coordinates.
+     *
+     * In follow-up work, this should be used to build & to update
+     * cleaner, public-facing segment objects for the touch path.
+     */
+    private _protoSegments: CumulativePathStats[] = [];
+
+    /**
+     * TODO: These directly represent the subsegments comprising their
+     * corresponding path segments (in `_protoSegments`) produced by the
+     * prototype algorithm.  Just... the stats analysis of each subsegment.
+     *
+     * In follow-up work, these should be used internally to build & update
+     * the internals of TBD cleaner, public-facing segment objects for the
+     * touch path.
+     */
+    private _protoSegmentSets: CumulativePathStats[][] = [];
+
+    /**
+     * Used to 'repeat' the most-recently observed incoming sample if no
+     * other replaces it before it triggers.
+     *
+     * A repeating sample indicates lack of motion, which is valuable
+     * information for stats-based segmentation.
+     */
+    private repeatTimer: number | NodeJS.Timeout;
```

Comment:
`NodeJS.Timeout` -- is this going to be a problem in web?

Reply:
Not at all.  In Web, it uses the `number` part, while in Node, it uses the `NodeJS.Timeout`.  `setTimeout`'s signature is different between the two.

- https://developer.mozilla.org/en-US/docs/Web/API/setTimeout
- https://nodejs.org/en/docs/guides/timers-in-node/#leaving-timeouts-behind

We don't care about the Node object's properties; all we care about is its use as a "handle" for timeout-related functions.  That aspect is 100% compatible with DOM's use of a number for the "handle" instead.

Reply:
Gotcha, it's a type definition, so goes away on JS side.